### ISSUE TITLE
Parser error with certain objects in top level variables in Chrome

### ISSUE THREAD

**Issue by Potato42** [Khan/live-editor#423]
I'm not entirely sure if this is meant as a security feature or not, but if you try to put references to `window` or the global program context into top level variables you get an elusive "Parser error" which happens only if you are using Chrome **and** have *any* PJS callback function defined anywhere in your program.
[This program](https://www.khanacademy.org/computer-programming/parser-error-in-chrome/6537414473547776) demonstrates the problem.  If you remove the draw function or use Firefox it works.  You can change the `win` definition to 

    var pjs = function() {return this;};
    pjs = pjs();

if you want to see the same problem with the program context instead of `window`.

So far, I have been unable to find the source of this bug.

**Comment by MikaalSky** [comment]
Yeah, that is strange. Just having the function be an anonymous function works fine, though, so IDK why you would set it as itself but the callback.

**Comment by Potato42** [comment]
@MikaalSky I can't quite understand what you are trying to say at the end there, but if you are talking about the `pjs` definition (which gets the global program context), it's different from the `win` definition (which gets `window`).  Some esprima stuff turns 

    var pjs = function() {return this;};

into

    __env__.pjs = function() {return this;};

So `pjs == __env__`.  My purpose for writing the `pjs` example is to show that it's not only the window object that gets messed up like that.  So far, these are the only two objects that I've seen act like that though. 

Is this what you were talking about in your comment?  If so, does it make sense why I did that now?

**Comment by Potato42** [comment]
I am now getting this error in both Firefox and Chrome.

**Comment by kevinbarabash** [comment]
@Potato42 it is weird that this causes a parse error, but it's low priority since returning this from an unbound method is either a mistake or someone is trying to escape the sandbox.

**Comment by kevinbarabash** [comment]
The issue is caused by the fact that we try to stringy all values that are objects so that we can do live updates if the user changes their data.  In this case, trying to serialize `window` doesn't work because it's a cyclic data structure.

**Comment by Potato42** [comment]
I had considered that earlier, but when I tried doing something like

    var cyclic = {};
    cyclic.cyclic = cyclic;

I never got an error.  Why is that?  Shouldn't any cyclic structure have the same result?  I'd play with it some more, but I seem to be unable to reproduce the error with anything right now.

**Comment by Potato42** [comment]
This is the weirdest thing.  I don't get the error any more, but the iframe now **reloads** every time you press the restart button, changing `document.referrer` to the kasandbox domain.  Usually restarting a program doesn't reload the iframe, but for some reason it does when you have the `draw` loop + `window`/`__env__` in a global variable combo.  Why on earth would that happen?
Plus, the fix I've been using for my evil escaping of the sandbox just wraps the return value in an object, so that's another reason I find the cyclic object explanation lacking.

**Comment by Willard21** [comment]
I was just checking to see if this had been reported. Another workaround to prevent the page reloading is to set all your callbacks (draw) directly to the pjs instance, rather than defining them normally.

This seems to have been fixed for the pjs version of the bug (a global pjs reference isn't buggy at all), but the window version is definitely still broken. It was the source of a lot of confusion when I started trying to make my 2,000+ line program maintain its state on edits, but found that the whole window was destroyed/reloaded each edit.
